### ISSUE TITLE
Merge argument order looks wrong in configuration_loader.lua

### ISSUE THREAD

**Issue by geoffleyland** [Olivine-Labs/busted#356]
tablex.merge preserves values from the table on the right when there's a conflict:

    $ lua
    > t = require"pl.tablex"
    > a=t.merge({a=1},{a=3},true)
    > =a.a
    3

But lines 17 and 25 of configuration_loader.lua have the default configuration on the right, and the config file on the left, so I suspect the config file can't overwrite anything.  Swapping them around fixed my problem anyway.

Looking further, it would appear that fixing #327 caused this.  As far as I can tell, at this point, config contains both default values and command-line values.  So it might be more a question of the order in which default, .busted and command-line values are merged, rather that swapping the order here.

Unfortunately I'm not sure you can tell the difference between what cliargs really read on the command line, and what it added for you as a default.

**Comment by o-lim** [comment]
I don't think the configuration file is supposed to overwrite anything.  The way I understand it is that, the command line is supposed to overwrite what is in the configuration file.

Quoting from http://olivinelabs.com/busted:
> Busted 1.6 added a concept of "tasks", or predefined busted configuration options. You can create a .busted file in the root, which is automatically loaded if it exists. Default options are run if no task is specified. You can add any argument available in the CLI (listed above), using the long name. (Use verbose = true, not v = true.) Any arguments you specify will override those set in the task.

**Comment by geoffleyland** [comment]
Yes, but the default arguments come through at the same time as the commands that were actually put on the command line.

For example, runner.lua has

    local lpathprefix = './src/?.lua;./src/?/?.lua;./src/?/init.lua'

Then

    cli:add_option('-m, --lpath=PATH', 'optional path to be prefixed to the Lua module search path', lpathprefix)

So by the time you get to

    local config, err = configLoader(configFile, cliArgs)

even if you only type `busted` on the command line, `cliArgs` contains a value for `lpath`.

So if you have a `.busted` with

    return { default = { lpath = "./my_path/?/?.lua" } }

It'll be ignored. (Try it!)

The problem isn't that things that are actually on the command-line are overridden, it's that the default values look like they came from the command-line.

**Comment by DorianGray** [comment]
tl;dr; If there is a .busted file, we need to -not- default cliargs.

**Comment by geoffleyland** [comment]
Is it as simple as that?  You might want some the other default args that are neither on the cli nor in .busted, and also, .busted is searched for in --cwd.
The only solution I've managed to come up with is having cli_args tell you which args are default and which were actually specified on the command line, and that seems hard enough that there must be a better solution.

**Comment by DorianGray** [comment]
Err, sorry, you're right, we do need field level control on defaults. Load the busted file and populate the values there as the defaults in cliargs.

**Comment by DorianGray** [comment]
Or, if they are not specified in .busted, use the default in code.

**Comment by o-lim** [comment]
@geoffleyland: Sorry, I didn't quite understand the issue at first, but I now I understand what the issue is.

It might be as simple as having a a `cliDefaultArgs` variable populated with the default values then pre-merge this with the `cliArgs`, returned from `cli:parse`, before sending to the configuration loader.

It woud look something like this:
```lua
local cliDefaultArgs = { ... }
...
local cliArgs = cli:parse(arg)
...
if bustedConfigFile then
  local config, err = configLoader(configFile, tablex.merge(cliDefaultArgs, cliArgs, true))
  ...
end
...
```
The problem is that `cliargs` sets all default values to `''` if the default is not provided, and flags are always defaulted to `false`.  However, a recent update to `cliargs` allows us to capture options as they are parsed by `cliargs`, making something like the above possible.

**Comment by DorianGray** [comment]
Why can't we reverse this and merge the config file with the built in defaults, then pass those values in as the defaults in cliargs?

**Comment by DorianGray** [comment]
...crap, I know why. We can specify the config file at the command line. We'd have to repeat the process twice if a custom one was specified.

**Comment by geoffleyland** [comment]
Another issue is that if you go `busted --help` then cli_args shows you the defaults, which is nice.  Any fix which involves not passing the defaults into cli_args would lose the "feature" of having the defaults displayed in the help.

**Comment by o-lim** [comment]
Probably need to pass the builtin defaults, cliArgs, and config file to the config loader.
```lua
local config, err = configLoader(configFile, cliDefaultArgs, cliArgs)
```
Then in configLoader do something like:
```lua
function configLoader(configFile, configDefaults, config)
  local runConfig = configFile[config.run]
  config = tablex.merge(runConfig, config, true)
  config = tablex.merge(configDefaults, config, true)
  return config
end
```
This assumes that `cliArgs` has not been populated with any default values.  So if it was not specified on the command-line, then it is `nil`.  This can be easily done with the latest version of `cliargs`, and still have the defaults passed to `cli:add_option`, preserving the defaults in the help message.

I'll see if I can whip up something later tonight.

**Comment by o-lim** [comment]
I made some changes that address this issue (o-lim/busted@e457fcc).

Let me know if this solves the problem or if there is still something missing.  If everything looks okay, I'll create a PR.

**Comment by geoffleyland** [comment]
This works for my test case, so as far as I'm concerned it's fixed.  Thank you!
